uast/original.go

```diff
@@ -0,0 +1,241 @@
+package uast
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+
+	"srcd.works/go-errors.v0"
+)
+
+var (
+	ErrEmptyAST             = errors.NewKind("input AST was empty")
+	ErrUnexpectedObject     = errors.NewKind("expected object of type %s, got: %#v")
+	ErrUnexpectedObjectSize = errors.NewKind("expected object of size %d, got %d")
+	ErrUnsupported          = errors.NewKind("unsupported: %s")
+)
+
+// ToNoder is a converter of source ASTs to *Node.
+type ToNoder interface {
+	// ToNode converts the source AST to a *Node.
+	ToNode(src interface{}) (*Node, error)
+}
+
+const (
+	// topLevelIsRootNode is true if the top level object is the root node
+	// of the AST. If false, top level object should have a single key, that
+	// being the root node.
+	topLevelIsRootNode = false
+	// InternalRoleKey is a key string uses in properties to use the internal
+	// role of a node in the AST, if any.
+	InternalRoleKey = "internalRole"
+)
+
+// BaseOriginalToNoder is an implementation of OriginalToNoder that aims to work
+// for the most common source ASTs.
+type BaseToNoder struct {
+	// InternalTypeKey is a key in the source AST that can be used to get the
+	// InternalType of a node.
+	InternalTypeKey string
+	// OffsetKey is a key that indicates the position offset.
+	OffsetKey string
+	// LineKey is a key that indicates the line number.
+	LineKey string
+	// TokenKeys is a slice of keys used to extract token content.
+	TokenKeys map[string]bool
+	// SyntheticTokens is a map of InternalType to string used to add
+	// synthetic tokens to nodes depending on its InternalType.
+	SyntheticTokens map[string]string
+}
+
+func (c *BaseToNoder) ToNode(v interface{}) (*Node, error) {
+	src, ok := v.(map[string]interface{})
+	if !ok {
+		return nil, ErrUnsupported.New("non-object root node")
+	}
+
+	if len(src) == 0 {
+		return nil, ErrEmptyAST.New()
+	}
+
+	if topLevelIsRootNode {
+		return nil, ErrUnsupported.New("top level object as root node")
+	}
+
+	if len(src) > 1 {
+		return nil, ErrUnexpectedObjectSize.New(1, len(src))
+	}
+
+	for _, obj := range src {
+		return c.toNode(obj)
+	}
+
+	panic("not reachable")
+}
+
+func (c *BaseToNoder) toNode(obj interface{}) (*Node, error) {
+	m, ok := obj.(map[string]interface{})
+	if !ok {
+		return nil, ErrUnexpectedObject.New("map[string]interface{}", obj)
+	}
+
+	n := NewNode()
+	for k, o := range m {
+
+		switch ov := o.(type) {
+		case map[string]interface{}:
+			child, err := c.mapToNode(k, ov)
+			if err != nil {
+				return nil, err
+			}
+
+			n.Children = append(n.Children, child)
+		case []interface{}:
+			children, err := c.sliceToNodes(k, ov)
+			if err != nil {
+				return nil, err
+			}
+
+			n.Children = append(n.Children, children...)
+		default:
+			if err := c.addProperty(n, k, o); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	sort.Sort(byOffset(n.Children))
```

Comment:
If we want the sorts to be always the same, we should use `sort.Stable` instead of `sort.Sort`.